### PR TITLE
Thing info changes are not handled

### PR DESCRIPTION
**Pull request by gboyvalenkov-bosch**
[#22] Thing info changes are not handled

- bug fixed

Signed-off-by: Georgi Boyvalenkov <Georgi.Boyvalenkov@bosch.io>

### CODE REVIEW

**Comment by e-grigorov** [comment]
@gboyvalenkov-bosch The Eclipse Kanto project is currently restructuring its license scheme - from Eclipse Public License 2.0 only to dual licensing under Eclipse Public License 2.0 **and** Apache License 2.0 (you can read more about it [here](https://projects.eclipse.org/projects/iot.kanto/reviews/restructuring-review)).

This contribution will be merged into the project with this new licensing scheme so its important for the project and the community to have your agreement on proceeding with it given these new circumstances :blush: 

Please, provide your feedback as a comment here that you, as an author of this contribution, understand and agree with the change explained above. Thank you!

**Comment by gboyvalenkov-bosch** [comment]
> @gboyvalenkov-bosch The Eclipse Kanto project is currently restructuring its license scheme - from Eclipse Public License 2.0 only to dual licensing under Eclipse Public License 2.0 **and** Apache License 2.0 (you can read more about it [here](https://projects.eclipse.org/projects/iot.kanto/reviews/restructuring-review)).
> 
> This contribution will be merged into the project with this new licensing scheme so its important for the project and the community to have your agreement on proceeding with it given these new circumstances 😊
> 
> Please, provide your feedback as a comment here that you, as an author of this contribution, understand and agree with the change explained above. Thank you!

I understand and agree with the license change explained above.

File: internal/feature.go [hunk 1]
```diff
@@ -85,26 +86,43 @@ func NewScriptBasedSU(scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig) (*Scr
...
+	logger.Infof("Connecting to ditto endpoint with configuration - %v", edgeCfg)
+	f.mqttClient = client
+	done = make(chan struct{})
+	f.queue = make(chan operationFunc, 10)
+	f.lock.Lock()
+	connected = true
+	f.lock.Unlock()
+	err := f.init(scriptSUPConfig, edgeCfg)
+	if err != nil {
+		return err
```
Comment (e-grigorov) [inline internal/feature.go:108]: In case of error, it'll stay connected.

File: internal/feature.go [hunk 2]
```diff
@@ -85,26 +86,43 @@ func NewScriptBasedSU(scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig) (*Scr
...
 // Connect the client to the configured Ditto endpoint.
 // If any error occurs during the connection's initiation - it's returned here.
-func (f *ScriptBasedSoftwareUpdatable) Connect() error {
+func (f *ScriptBasedSoftwareUpdatable) Connect(client MQTT.Client, scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig, edgeCfg *edgeConfiguration) error {
+	logger.Infof("Connecting to ditto endpoint with configuration - %v", edgeCfg)
+	f.mqttClient = client
+	done = make(chan struct{})
+	f.queue = make(chan operationFunc, 10)
+	f.lock.Lock()
+	connected = true
```
Comment (e-grigorov) [inline internal/feature.go:104]: It seems like a race condition. the flag will indicate that the feature is connected but it's not. Connect and disconnect has to be in synch in the context of their initialization.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/edge.go [hunk 3]
```diff
@@ -32,55 +31,79 @@ type edgeConfiguration struct {
 	PolicyID string `json:"policyId"`
 }
 
-var edgeConfigurationChan = make(chan *edgeConfiguration, 1)
+// EdgeConnector listens for Edge Thing configuration changes and notifies the corresponding EdgeClient
+type EdgeConnector struct {
```
[the remark this answers is not in the document]
Comment (e-grigorov) [reply internal/edge.go:35]: Actually, it's used in the main package.

File: internal/edge.go [hunk 4]
```diff
@@ -32,55 +31,79 @@ type edgeConfiguration struct {
...
+// edgeClient receives notifications of Edge Thing configuration changes from EdgeConnector
+type edgeClient interface {
+	Connect(client MQTT.Client, scriptSUPConfig *ScriptBasedSoftwareUpdatableConfig, cfg *edgeConfiguration) error
+	Disconnect(closeStorage bool)
+}
 
-// retrieveEdgeConfiguration tries to connect to the local MQTT broker and retrieve its configuration.
-func retrieveEdgeConfiguration(server string, username string, password string) (*edgeConfiguration, MQTT.Client, error) {
-	logger.Infof("Retrieve edge configuration from: %s", server)
+// newEdgeConnector create EdgeConnector with the given server, username and password for the given EdgeClient
```
Comment (e-grigorov) [inline internal/edge.go:47]: - create -> creates
- with the given server, username and password -> with the given configuration

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/edge.go [hunk 5]
```diff
@@ -32,55 +31,80 @@ type edgeConfiguration struct {
...
+			logger.Infof("apply edge configuration: %v", localCfg)
+			if p.cfg != nil {
+				p.edgeClient.Disconnect(false)
+			}
+			p.cfg = localCfg
+			err = ecl.Connect(p.mqttClient, scriptSUPConfig, p.cfg)
+			if err != nil {
+				logger.Errorf("error connecting to broker: %v", err)
+			} else {
+				logger.Infof("Edge [TenantID: %s, DeviceID: %s, PolicyID: %s]", p.cfg.TenantID, p.cfg.DeviceID, p.cfg.PolicyID)
```
Comment (e-grigorov) [inline internal/edge.go:84]: Edge -> edge

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/feature_internal.go [hunk 6]
```diff
@@ -98,7 +98,13 @@ func (f *ScriptBasedSoftwareUpdatable) load() {
 }
 
 func (f *ScriptBasedSoftwareUpdatable) process() {
+	f.lock.Lock()
+	if !featureAvailable {
+		f.lock.Unlock()
+		return
```
Comment (e-grigorov) [inline internal/feature_internal.go:104]: If edge client is not connected, the processing will not start. The problematic scenario:
- run software-update and connect to the Mosquitto broker
- after that run suite-connect

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/feature_internal.go (hunk 6 above)
Comment (gboyvalenkov-bosch) [reply internal/feature_internal.go:104]: Checked, method is only called after suite-connect

Comment (e-grigorov) [reply internal/feature_internal.go:104]: There is execution flow: paho for thing info -> ditto client to process. No issue at all.

**Pull request merged**